pkg/oauth/api/types.go

```diff
@@ -73,6 +73,34 @@ type OAuthClient struct {
 
 	// RedirectURIs is the valid redirection URIs associated with a client
 	RedirectURIs []string
+
+	// ScopeRestrictions describes which scopes this client can request.  Each requested scope
+	// is checked against each restriction.  If any restriction matches, then the scope is allowed.
+	// If no restriction matches, then the scope is denied.
+	ScopeRestrictions []ScopeRestriction
+
+	// UnrestrictedAccess indicates that the client is allowed to request a token unconstrained by scopes.
+	// If this is true, then ScopeRestrictions is ignored.
+	UnrestrictedAccess bool
+}
+
+// ScopeRestriction describe one restriction on scopes.  Exactly one option must be non-nil.
+type ScopeRestriction struct {
+	// Literals means the scope has to match a particular set of strings exactly
+	Literals []string
```

Comment:
I would prefer "Exact" or "ExactScopes" (although that stutters) or "ExactValues" because I'd rather not introduce a new word "literals" that while accurate, isn't associated with this concept space.